File: ui/src/lib/seasonCommissioner.ts
// SeasonCommissioner API client. A SeasonCommissioner links a Season to a User
// acting as a co-commissioner. The read surface is the generic CRUD registered
// in main.go; writes go through the custom routes in
// route/seasoncommissioner (same paths), which enforce the model's sysadmin-only
// EditableBy constraint:
//
//	GET    /api/season_commissioner      -> { resource: SeasonCommissioner[] }
//	GET    /api/season_commissioner/:id  -> { resource: SeasonCommissioner }
//	POST   /api/season_commissioner      -> { resource: SeasonCommissioner }
//	DELETE /api/season_commissioner/:id  -> { resource: SeasonCommissioner }
//
// Records are readable by everyone; only a system administrator may create or
// delete them (see model.SeasonCommissioner.EditableBy). Record IDs are
// 16-char hex strings.
import { authFetch } from '$lib/auth';

export interface SeasonCommissioner {
	id: string;
	season_id: string;
	user_id: string;
	created_at: string;
	updated_at: string;
}

const BASE = '/api/season_commissioner';

// unwrap parses a CrudCommon response (`{ resource: ... }`) and throws the
// backend error message on non-2xx responses so callers can surface it.
async function unwrap(res: Response): Promise<any> {
	if (!res.ok) {
		let message = `Request failed (${res.status})`;
		try {
			const body = await res.json();
			if (body?.error) message = body.error;
		} catch {
			// non-JSON error body; fall back to the status message
		}
		throw new Error(message);
	}
	const body = await res.json();
	return body?.resource;
}

export async function listSeasonCommissioners(): Promise<SeasonCommissioner[]> {
	return unwrap(await authFetch(BASE));
}

// addSeasonCommissioner adds a user to a season as a co-commissioner.
export async function addSeasonCommissioner(seasonId: string, userId: string): Promise<SeasonCommissioner> {
	return unwrap(
		await authFetch(BASE, {
			method: 'POST',
			headers: { 'Content-Type': 'application/json' },
			body: JSON.stringify({ season_id: seasonId, user_id: userId })
		})
	);
}

export async function removeSeasonCommissioner(id: string): Promise<void> {
	await authFetch(`${BASE}/${id}`, { method: 'DELETE' });
}

File: route/seasoncommissioner/season_commissioner.go
// Package seasoncommissioner exposes the write surface for SeasonCommissioner
// records (adding/removing co-commissioners to a season).
//
// The generic CRUD surface in main.go is registered read-only for this type;
// writes go exclusively through the routes in this package, which enforce the
// model's sysadmin-only EditableBy constraint.
package seasoncommissioner

import (
	"context"
	"errors"
	"net/http"

	"intraclub/api"
	"intraclub/database"
	"intraclub/model"

	"github.com/gin-gonic/gin"
)

// BaseRoute is the base API route for SeasonCommissioner records. It matches
// the model's Type() ("season_commissioner").
var BaseRoute = "/season_commissioner"

// AddCommissionerBody is the request body for AddCommissioner.
type AddCommissionerBody struct {
	// SeasonId is the season the user is being added as a co-commissioner to.
	SeasonId model.SeasonId `json:"season_id"`

	// UserId is the user being added as a co-commissioner.
	UserId database.UserId `json:"user_id"`
}

// StaticallyValid ensures both a season and a user were provided.
func (b *AddCommissionerBody) StaticallyValid() error {
	if b.SeasonId == model.SeasonId(database.InvalidRecordId) {
		return errors.New("season_id must not be empty")
	}
	if b.UserId == database.InvalidUserId {
		return errors.New("user_id must not be empty")
	}
	return nil
}

// AddCommissioner creates a SeasonCommissioner record linking the given user
// to the given season. Only a system administrator may add co-commissioners.
type AddCommissioner struct{}

func (c AddCommissioner) Path() (api.HttpMethod, string) {
	return api.HttpMethodPost, BaseRoute
}

func (c AddCommissioner) RequestBody() (*AddCommissionerBody, bool) {
	return &AddCommissionerBody{}, true
}

func (c AddCommissioner) Handler(req api.Request[*AddCommissionerBody]) (any, int, error) {
	if req.Token == nil {
		return nil, http.StatusUnauthorized, errors.New("token is required")
	}

	isAdmin, err := model.IsUserSystemAdministrator(req.Context, req.DatabaseProvider, req.Token.UserId)
	if err != nil {
		return nil, http.StatusInternalServerError, err
	}
	if !isAdmin {
		return nil, http.StatusForbidden, errors.New("only a system administrator may add co-commissioners to a season")
	}

	season, err := database.GetExistingRecordById(req.Context, req.DatabaseProvider, &model.Season{}, req.Body.SeasonId.RecordId())
	if err != nil {
		return nil, http.StatusBadRequest, err
	}
	if err := season.AddCommissioner(req.Context, req.DatabaseProvider, req.Body.UserId); err != nil {
		return nil, http.StatusBadRequest, err
	}

	// Return the created record so the client can use its ID to remove it.
	commissioners, err := database.GetAllWhere[*model.SeasonCommissioner](req.Context, req.DatabaseProvider,
		func(_ context.Context, c *model.SeasonCommissioner) bool {
			return c.SeasonId == season.ID && c.UserId == req.Body.UserId
		})
	if err != nil {
		return nil, http.StatusInternalServerError, err
	}
	if len(commissioners) == 0 {
		return nil, http.StatusInternalServerError, errors.New("co-commissioner was not created")
	}
	return gin.H{api.ResourceKey: commissioners[0]}, http.StatusOK, nil
}

// RemoveCommissioner deletes the SeasonCommissioner record with the given ID.
// Only a system administrator may remove co-commissioners.
type RemoveCommissioner struct{}

func (c RemoveCommissioner) Path() (api.HttpMethod, string) {
	return api.HttpMethodDelete, api.AppendPathId(BaseRoute)
}

func (c RemoveCommissioner) RequestBody() (*AddCommissionerBody, bool) {
	return &AddCommissionerBody{}, false
}

func (c RemoveCommissioner) Handler(req api.Request[*AddCommissionerBody]) (any, int, error) {
	if req.Token == nil {
		return nil, http.StatusUnauthorized, errors.New("token is required")
	}

	isAdmin, err := model.IsUserSystemAdministrator(req.Context, req.DatabaseProvider, req.Token.UserId)
	if err != nil {
		return nil, http.StatusInternalServerError, err
	}
	if !isAdmin {
		return nil, http.StatusForbidden, errors.New("only a system administrator may remove co-commissioners from a season")
	}

	existing, exists, err := database.GetOneById(req.Context, req.DatabaseProvider, &model.SeasonCommissioner{}, req.PathId)
	if err != nil {
		return nil, http.StatusInternalServerError, err
	}
	if !exists {
		return nil, http.StatusNotFound, errors.New("co-commissioner not found")
	}
	if _, _, err := database.DeleteOneById(req.Context, req.DatabaseProvider, existing, existing.ID); err != nil {
		return nil, http.StatusInternalServerError, err
	}
	return gin.H{api.ResourceKey: existing}, http.StatusOK, nil
}
